File: productpulse-ai/src/services/productService.js
import * as productsApi from '../api/products'
import { toServiceError } from '../api/axios'

export async function fetchProducts(params) {
  try {
    const response = await productsApi.listProducts(params)
    return response.data
  } catch (error) {
    throw toServiceError(error)
  }
}

export async function fetchProduct(id) {
  try {
    const response = await productsApi.getProduct(id)
    return response.data
  } catch (error) {
    throw toServiceError(error)
  }
}

/** @param {{ name: string, description?: string }} payload */
export async function createProduct(payload) {
  try {
    const response = await productsApi.createProduct(payload)
    return response.data
  } catch (error) {
    throw toServiceError(error)
  }
}

/** @param {{ name?: string, description?: string }} payload */
export async function updateProduct(id, payload) {
  try {
    const response = await productsApi.updateProduct(id, payload)
    return response.data
  } catch (error) {
    throw toServiceError(error)
  }
}

export async function deleteProduct(id) {
  try {
    await productsApi.deleteProduct(id)
  } catch (error) {
    throw toServiceError(error)
  }
}

File: productpulse-backend/app/services/product_service.py
import uuid

from sqlalchemy.orm import Session

from app.core.exceptions import NotFoundException
from app.models.product import Product
from app.repositories import product_repository
from app.schemas.product import ProductCreate, ProductUpdate


def list_products(db: Session, user_id: uuid.UUID, *, page: int, limit: int, search: str | None) -> tuple[list[Product], int]:
    return product_repository.list_for_user(db, user_id, page=page, limit=limit, search=search)


def get_product(db: Session, product_id: uuid.UUID, user_id: uuid.UUID) -> Product:
    product = product_repository.get_by_id_for_user(db, product_id, user_id)
    if not product:
        raise NotFoundException("Product not found.", error_code="PRODUCT_NOT_FOUND")
    return product


def create_product(db: Session, payload: ProductCreate, user_id: uuid.UUID) -> Product:
    return product_repository.create(db, user_id=user_id, name=payload.name, description=payload.description)


def update_product(db: Session, product_id: uuid.UUID, payload: ProductUpdate, user_id: uuid.UUID) -> Product:
    product = get_product(db, product_id, user_id)  # raises NotFoundException if missing/not owned
    return product_repository.update(db, product, name=payload.name, description=payload.description)


def delete_product(db: Session, product_id: uuid.UUID, user_id: uuid.UUID) -> None:
    product = get_product(db, product_id, user_id)
    product_repository.delete(db, product)
